pkg/controller/component/vars_test.go

```diff
@@ -2238,5 +2242,326 @@ var _ = Describe("vars", func() {
+			})
+
+			It("exp for credential-vars", func() {
+				vars := []appsv1alpha1.EnvVar{
+					{
+						Name: "password",
+						ValueFrom: &appsv1alpha1.VarSource{
+							CredentialVarRef: &appsv1alpha1.CredentialVarSelector{
+								ClusterObjectReference: appsv1alpha1.ClusterObjectReference{
+									Name:     "credential",
+									Optional: required(),
+								},
+								CredentialVars: appsv1alpha1.CredentialVars{
+									Password: &appsv1alpha1.VarRequired,
+								},
+							},
+						},
+						Expression: expp("panic"), // the expression will not be evaluated
+					},
+				}
+				reader := &mockReader{
+					cli: testCtx.Cli,
+					objs: []client.Object{
+						&corev1.Secret{
+							ObjectMeta: metav1.ObjectMeta{
+								Namespace: testCtx.DefaultNamespace,
+								Name:      constant.GenerateAccountSecretName(synthesizedComp.ClusterName, synthesizedComp.Name, "credential"),
+							},
+							Data: map[string][]byte{
+								constant.AccountPasswdForSecret: []byte("password"),
+							},
+						},
+					},
+				}
+				templateVars, envVars, err := ResolveTemplateNEnvVars(testCtx.Ctx, reader, synthesizedComp, vars)
+				Expect(err).Should(Succeed())
+				Expect(templateVars).ShouldNot(HaveKey("password"))
+				checkEnvVarWithValueFrom(envVars, "password", &corev1.EnvVarSource{
+					SecretKeyRef: &corev1.SecretKeySelector{
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: reader.objs[0].GetName(),
+						},
+						Key: constant.AccountPasswdForSecret,
+					},
+				})
+			})
+
+			It("depends on credential-vars", func() {
+				vars := []appsv1alpha1.EnvVar{
+					{
+						Name: "raw",
+						ValueFrom: &appsv1alpha1.VarSource{
+							CredentialVarRef: &appsv1alpha1.CredentialVarSelector{
+								ClusterObjectReference: appsv1alpha1.ClusterObjectReference{
+									Name:     "credential",
+									Optional: required(),
+								},
+								CredentialVars: appsv1alpha1.CredentialVars{
+									Password: &appsv1alpha1.VarRequired,
+								},
+							},
+						},
+						Expression: expp("panic"),
+					},
+					{
+						Name:       "password",
+						Expression: expp("{{ .raw }}"), // depends on $raw which is a credential-var
+					},
+				}
+				reader := &mockReader{
+					cli: testCtx.Cli,
+					objs: []client.Object{
+						&corev1.Secret{
+							ObjectMeta: metav1.ObjectMeta{
+								Namespace: testCtx.DefaultNamespace,
+								Name:      constant.GenerateAccountSecretName(synthesizedComp.ClusterName, synthesizedComp.Name, "credential"),
+							},
+							Data: map[string][]byte{
+								constant.AccountNameForSecret:   []byte("username"),
+								constant.AccountPasswdForSecret: []byte("password"),
+							},
+						},
+					},
+				}
+				_, _, err := ResolveTemplateNEnvVars(testCtx.Ctx, reader, synthesizedComp, vars)
+				Expect(err).ShouldNot(Succeed())
+				Expect(err.Error()).Should(And(ContainSubstring("map has no entry for key"), ContainSubstring("raw")))
```

Comment:
Why expression doesn't work for credential vars?

Reply:
This is prohibited by design, to protect against the transmission of credentials in plain text within the system.